woocommerce/woocommerce-ios
# [Mobile Payments] Implement simpler built-in reader connection flow

Part of: #8082


## Description


In sTAP Away, we add the ability to use the built-in NFC card reader on compatible iPhones to take card payments.

The Stripe SDK we use keeps a very similar interface for built in reader connections, compared to bluetooth reader connections. The interface is almost identical, asking us to discover a reader, then connect to it, even though it's the phone's built in hardware.

The connection flows for BuiltIn and Bluetooth card readers are similar, but the built-in flow has less complexity. In particular, we only need to support one built-in reader for Tap on Mobile, so we can remove the ability to skip a reader, don't need to show a list of readers to choose between, and can always auto connect to a found reader.

Using the BuiltIn variant of the CardReaderConnectionController (duplicated in #8280) I've stripped out the unnecessary steps and code to provide the following connection flow.

1. Clear any `candidateReader`
2. Start a `localMobile` search using `CardPresentPaymentAction.startCardReaderDiscovery`
3. When the first reader is found, start connecting to it
4. If there's a "software update" (actually configuring the device), install it
5. When the reader is connected, call completion with `connected(reader)` so that Preflight can continue with its work
6. Display any errors, allowing retry where appropriate.

**N.B. There's no change to the visuals or text for the Built-in reader flow** – this PR continues to use the same alerts as the bluetooth reader flow.

The next step is to update the alerts that are shown, and then, I'll factor out the shared code.

## Testing instructions


If you do not have a device which is provisioned for Tap on Mobile, please enable the simulated card reader and build to a simulator. 

To enable the Stripe simulated reader, in Xcode go to `Product > Scheme > Edit Scheme` and select `Run` in the sidebar, then `Arguments` in the tabs. Check the `-simulate-stripe-card-reader` box.

Using a US store with WCPay or the Stripe gateway:

1. Turn on the `Tap to Pay on iPhone` experimental feature in settings
2. Tap `Menu > Payments > Collect Payment`
3. Follow the flow to create a payment, and select Card when asked for the payment method
4. Choose Tap to Pay on iPhone
5. Observe that the connection flow proceeds and automatically connects to the built in reader.

Repeat, using the simulated reader, [after setting the software update type to](https://github.com/woocommerce/woocommerce-ios/blob/db5863f26f158e47813dfcfa8f4de6d9702a44ef/Hardware/Hardware/CardReader/StripeCardReader/StripeCardReaderService.swift#L78) `.required` (change it in code and rebuild)

Observe that when you do this, the connection flow proceeds and performs the update, displaying progress as it does so.

## Screenshots



https://user-images.githubusercontent.com/2472348/205089145-dd191cb1-adf0-49ae-abad-aaedaa595d15.mp4



---
